crossdock/client/tchclient/thrift.go

```diff
@@ -39,21 +44,377 @@ func runThrift(t crossdock.T, call call) {
+					-2: true,
+				},
+			},
+			Want: map[int32]bool{
+				1:  true,
+				2:  true,
+				-1: true,
+				-2: true,
+			},
+		},
+		{
+			Function: "TestString",
+			Give:     []interface{}{token},
+			Want:     token,
+		},
+		{
+			Function: "TestStringMap",
+			Give: []interface{}{
+				map[string]string{
+					"foo":   "bar",
+					"hello": "world",
+				},
+			},
+			Want: map[string]string{
+				"foo":   "bar",
+				"hello": "world",
+			},
+		},
+		{
+			Function: "TestStruct",
+			Give: []interface{}{
+				&gauntlet_apache.Xtruct{
+					StringThing: ptr.String("0"),
+					ByteThing:   ptr.Byte(1),
+					I32Thing:    ptr.Int32(2),
+					I64Thing:    ptr.Int64(3),
+				},
+			},
+			Want: &gauntlet_apache.Xtruct{
+				StringThing: ptr.String("0"),
+				ByteThing:   ptr.Byte(1),
+				I32Thing:    ptr.Int32(2),
+				I64Thing:    ptr.Int64(3),
+			},
+		},
+		{
+			Function: "TestTypedef",
+			Give:     []interface{}{gauntlet_apache.UserId(42)},
+			Want:     gauntlet_apache.UserId(42),
+		},
+		{
+			Function: "TestVoid",
+			Give:     []interface{}{},
+		},
+		{
+			Service:  "SecondService",
+			Function: "BlahBlah",
+			Give:     []interface{}{},
+			// TODO no Want here?
+		},
+		{
+			Service:  "SecondService",
+			Function: "SecondtestString",
+			Give:     []interface{}{"hello"},
+			Want:     "hello",
+		},
+	}
+
+	for _, tt := range tests {
+		desc := gauntlet.BuildDesc(tt)
+
+		client := buildClient(t, desc, tt.Service, thrift.NewClient(call.Channel, serverName, nil))
+		f := client.MethodByName(tt.Function)
+		if !checks.True(f.IsValid(), "%v: invalid function", desc) {
+			continue
+		}
+
+		ctx, cancel := thrift.NewContext(time.Second)
+		defer cancel()
+
+		args := []reflect.Value{reflect.ValueOf(ctx)}
+		if give, ok := gauntlet.BuildArgs(t, desc, f.Type(), tt.Give); ok {
+			args = append(args, give...)
+		} else {
+			continue
+		}
+
+		got, err := extractCallResponse(t, desc, f.Call(args))
+		if isUnrecognizedProcedure(err) {
+			t.Skipf("%v: procedure not defined", desc)
+			continue
+		}
+		gauntlet.Assert(t, tt, desc, got, err)
+	}
+}
+
+func buildClient(t crossdock.T, desc string, service string, client thrift.TChanClient) reflect.Value {
+	switch service {
+	case "", "ThriftTest":
+		client := gauntlet_apache.NewTChanThriftTestClient(client)
+		return reflect.ValueOf(client)
+	case "SecondService":
+		client := gauntlet_apache.NewTChanSecondServiceClient(client)
+		return reflect.ValueOf(client)
+	default:
+		crossdock.Fatals(t).Fail("", "%v: unknown thrift service", desc)
+		return reflect.Value{} // we'll never actually get here
+	}
+}
+
+// TODO once other servers implement the gauntlet, this func should be removed
+func isUnrecognizedProcedure(err error) bool {
+	if err == nil {
+		return false
+	}
+	return strings.Contains(err.Error(), "unrecognized procedure")
+}
+
+func extractCallResponse(t crossdock.T, desc string, returns []reflect.Value) (interface{}, error) {
+	var (
+		err error
+		got interface{}
+	)
```

Comment:
@abhinav @prashantv is it better to just used named return values here rather than a factored `var` declaration?